### ISSUE TITLE
Normalization layer masking

### ISSUE THREAD

**Issue by ianzur** [keras-team/keras#15451]
**System information**.

TensorFlow version (you are using): tf-nightly (on CoLab), locally 2.7.0 (built locally with ROCm support, a few weeks behind)
Are you willing to contribute it (Yes/No) : Yes, after maintainers evaluate proposal

**Describe the feature and the current behavior/state.**
tf.keras.layers.Normalization() does not ignore masked values.

**Will this change the current api? How?**
Only additional parameters, should not contain any non-backwards compatible changes.

**Who will benefit with this feature?**
People working with ragged data who want to include a preprocessing layer to normalize in model.

[Colab gist testing current behavior](https://colab.research.google.com/drive/1aKGeS0lJy-QEQx0jARyOAqAMFBEFRe--?usp=sharing)

- Do you want to contribute a PR? (yes/no): yes
- Briefly describe your candidate solution(if contributing):
  - add parameter to `.adapt()` method (mask_value=0.0 and/or boolean mask)
  - add `call(..., mask=None)` and tensor multiply to allow masking during model fitting/evaluation
 
An attempt at masked normalization (does **NOT** generalize to any input shape) 
```
class MaskedNormalization(tf.keras.layers.Layer):
    """custom masked normalization layer"""

    def __init__(
        self,
        mask_value: int = 999,
        input_dims: tuple = (51, 3),
        name: str = "masked_normalizer",
    ):
        super(MaskedNormalization, self).__init__()

        self.mask_value = mask_value
        self.supports_masking = True
 
        self.normalization = tf.keras.layers.experimental.preprocessing.Normalization(axis=2)


    def adapt(self, inputs):
        """redefined adapt to ignore masked values"""
        if not self.normalization.built:
            raise RuntimeError("layer must be \"built\" prior to calling `.adapt()`, see `model.build_graph()`")

        mask = inputs != self.mask_value

        float_mask = tf.cast(mask, dtype=tf.dtypes.float32)

        mean = tf.reduce_sum(inputs * float_mask, axis=[0, 1]) / tf.reduce_sum(
            float_mask, axis=[0, 1]
        )
        # sample variance (divide by: n-1)
        var = tf.reduce_sum(((inputs - mean) ** 2) * float_mask, axis=[0, 1]) / (
            tf.reduce_sum(float_mask, axis=[0, 1])
            - tf.constant(1, dtype=tf.dtypes.float32, shape=(3,))
        )
        count = tf.experimental.numpy.count_nonzero(
            float_mask
        )

        self.normalization.adapt_mean.assign(mean)
        self.normalization.adapt_variance.assign(var)
        self.normalization.count.assign(count)

        self.normalization.finalize_state()
        self.normalization._is_adapted = True

    def call(self, inputs, mask=None):

        # if mask provided mask output values (multiply by zero)
        if mask is not None:
                reshape_mask = tf.transpose(
                    tf.expand_dims(tf.cast(mask, dtype=tf.dtypes.float32), axis=0),
                    perm=[1, 2, 0],
                )
            return self.normalization(inputs) * reshape_mask

        return self.normalization(inputs)

    def compute_mask(self, inputs, mask=None):
        """return the input_mask directly"""
        return mask
```

**Comment by LukeWood** [comment]
Hey there!  We are interested in this feature, please go ahead and prepare a pull request if you are willing!

**Comment by MrinalTyagi** [comment]
I would like to work on this

**Comment by ianzur** [comment]
Go for it! This fell on my back burner

**Comment by kydonian** [comment]
Related if you're here looking for a workaround: created this version that plays nicely with `tf.data.Dataset`: https://gist.github.com/kydonian/ddda453c8707cabb1eb9fbcbbd335ffa

**Comment by SuryanarayanaY** [comment]
Hello, Thank you for reporting an issue.

We're currently in the process of migrating the new Keras 3 code base from keras-team/keras-core to keras-team/keras.
Consequently, This issue may not be relevant to the Keras 3 code base . After the migration is successfully completed, feel free to reopen this Issue at keras-team/keras if you believe it remains relevant to the Keras 3 code base. If instead this Issue is a bug or security issue in legacy tf.keras, you can instead report a new issue at [keras-team/tf-keras](https://github.com/keras-team/tf-keras/issues), which hosts the TensorFlow-only, legacy version of Keras.

To know more about Keras 3, please read https://keras.io/keras_core/announcement/
